In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from tabulate import tabulate

%matplotlib inline

In [2]:
INDEX_FILE = r'D:\petrtsv\projects\ds\few_shot_learning_study\index.csv'

records = pd.read_csv(INDEX_FILE, index_col=0)
records = records[
    (records['version'].isin({
        '1.0.6', '1.0.7', '1.0.8', '1.0.9', '1.0.10', '1.0.12'
    }))
]

records = records[records['is_test'] == True]
records = records[records['distance'] != 'sen']

records.columns

Index(['Name', 'Explanation', 'n_iterations', 'n_way', 'n_shot', 'batch_size',
       'eval_period', 'eval_iterations', 'time', 'accuracy', 'iteration',
       'result_directory', 'image_size', 'backbone', 'aux_location_k',
       'aux_rotation_k', 'dataset', 'train_classes', 'test_n_iterations',
       'dfmn_k', 'distance', 'k', 'test_batch_size', 'test_dataset',
       'test_n_way', 'test_accuracy', 'testing_time', 'accuracy_std',
       'top_3_accuracy', 'top_3_accuracy_std', 'test_accuracy_std',
       'test_top_3_accuracy', 'test_top_3_accuracy_std', 'version', 'is_test',
       'test_balanced', 'is_feat', 'is_relevance'],
      dtype='object')

In [3]:
COLUMNS = ['dataset', 'test_dataset', 'n_shot', 'k', 'aux_rotation_k', 'dfmn_k', 'is_feat',
           'distance', 'test_accuracy', 'test_accuracy_std', 'test_top_3_accuracy', 'test_top_3_accuracy_std']
records = records[COLUMNS].sort_values(by='test_accuracy', ascending=False)
# records.head()

# Raw resuts tables

* ## MiniImageNet -> MiniImageNet (5-way)

In [4]:
records_miniImageNet_miniImageNet = records[(records['dataset'] == 'miniImageNet') & (
    records['test_dataset'] == 'miniImageNet-test')]
records_miniImageNet_miniImageNet

dataset       test_dataset  n_shot    k  aux_rotation_k  dfmn_k  \
335  miniImageNet  miniImageNet-test     5.0  0.2             0.0     1.0   
326  miniImageNet  miniImageNet-test     5.0  1.0             1.0     0.0   
320  miniImageNet  miniImageNet-test     5.0  1.0             1.0     0.0   
329  miniImageNet  miniImageNet-test     5.0  0.2             0.0     1.0   
341  miniImageNet  miniImageNet-test     5.0  0.2             0.0     1.0   
293  miniImageNet  miniImageNet-test     5.0  1.0             0.0     0.0   
314  miniImageNet  miniImageNet-test     5.0  1.0             1.0     0.0   
344  miniImageNet  miniImageNet-test     5.0  0.2             0.0     1.0   
332  miniImageNet  miniImageNet-test     5.0  0.2             0.0     1.0   
347  miniImageNet  miniImageNet-test     5.0  0.2             1.0     1.0   
311  miniImageNet  miniImageNet-test     5.0  1.0             1.0     0.0   
323  miniImageNet  miniImageNet-test     5.0  1.0             1.0     0.0   
299  miniImageNet  miniImageNet-test     5.0  1.0             0.0     0.0   
338  miniImageNet  miniImageNet-test     5.0  0.2             0.0     1.0   
356  miniImageNet  miniImageNet-test     5.0  0.2             1.0     1.0   
362  miniImageNet  miniImageNet-test     5.0  0.2             1.0     1.0   
353  miniImageNet  miniImageNet-test     5.0  0.2             1.0     1.0   
296  miniImageNet  miniImageNet-test     5.0  1.0             0.0     0.0   
359  miniImageNet  miniImageNet-test     5.0  0.2             1.0     1.0   
317  miniImageNet  miniImageNet-test     5.0  1.0             1.0     0.0   
302  miniImageNet  miniImageNet-test     5.0  1.0             0.0     0.0   
290  miniImageNet  miniImageNet-test     5.0  1.0             0.0     0.0   
305  miniImageNet  miniImageNet-test     5.0  1.0             0.0     0.0   
350  miniImageNet  miniImageNet-test     5.0  0.2             1.0     1.0   
256  miniImageNet  miniImageNet-test     1.0  1.0             1.0     0.0   
262  miniImageNet  miniImageNet-test     1.0  1.0             1.0     0.0   
244  miniImageNet  miniImageNet-test     1.0  1.0             0.0     0.0   
268  miniImageNet  miniImageNet-test     1.0  0.2             0.0     1.0   
274  miniImageNet  miniImageNet-test     1.0  0.2             0.0     1.0   
259  miniImageNet  miniImageNet-test     1.0  1.0             1.0     0.0   
250  miniImageNet  miniImageNet-test     1.0  1.0             0.0     0.0   
247  miniImageNet  miniImageNet-test     1.0  1.0             0.0     0.0   
238  miniImageNet  miniImageNet-test     1.0  0.2             1.0     1.0   
102  miniImageNet  miniImageNet-test     1.0  0.2             0.0     1.0   
234  miniImageNet  miniImageNet-test     1.0  0.2             1.0     1.0   
271  miniImageNet  miniImageNet-test     1.0  0.2             0.0     1.0   
287  miniImageNet  miniImageNet-test     1.0  0.2             1.0     1.0   
105  miniImageNet  miniImageNet-test     1.0  0.2             0.0     1.0   
96   miniImageNet  miniImageNet-test     1.0  1.0             1.0     0.0   
111  miniImageNet  miniImageNet-test     1.0  0.2             1.0     1.0   
90   miniImageNet  miniImageNet-test     1.0  1.0             0.0     0.0   
265  miniImageNet  miniImageNet-test     1.0  0.2             0.0     1.0   
231  miniImageNet  miniImageNet-test     1.0  0.2             1.0     1.0   
253  miniImageNet  miniImageNet-test     1.0  1.0             1.0     0.0   
108  miniImageNet  miniImageNet-test     1.0  0.2             1.0     1.0   
99   miniImageNet  miniImageNet-test     1.0  1.0             1.0     0.0   
241  miniImageNet  miniImageNet-test     1.0  1.0             0.0     0.0   
93   miniImageNet  miniImageNet-test     1.0  1.0             0.0     0.0   

    is_feat      distance  test_accuracy  test_accuracy_std  \
335   False  cosine_scale        0.70256           0.011138   
326    True     euclidean        0.69976           0.011174   
320    True  cosine_scale        0.69976       

* ## GoogleLandmarks -> GoogleLandmarks (100-way)

In [5]:
records_googleLandmarks_googleLandmarks = records[(
    records['dataset'] == 'google-landmarks') & (records['test_dataset'] == 'google-landmarks-test')]
records_googleLandmarks_googleLandmarks

dataset           test_dataset  n_shot    k  aux_rotation_k  \
176  google-landmarks  google-landmarks-test     5.0  1.0             0.0   
154  google-landmarks  google-landmarks-test     5.0  0.2             0.0   
185  google-landmarks  google-landmarks-test     5.0  1.0             1.0   
221  google-landmarks  google-landmarks-test     5.0  1.0             1.0   
212  google-landmarks  google-landmarks-test     5.0  1.0             0.0   
215  google-landmarks  google-landmarks-test     5.0  1.0             0.0   
145  google-landmarks  google-landmarks-test     5.0  1.0             1.0   
218  google-landmarks  google-landmarks-test     5.0  1.0             1.0   
136  google-landmarks  google-landmarks-test     5.0  1.0             0.0   
182  google-landmarks  google-landmarks-test     5.0  1.0             1.0   
224  google-landmarks  google-landmarks-test     5.0  0.2             0.0   
200  google-landmarks  google-landmarks-test     5.0  0.2             1.0   
163  google-landmarks  google-landmarks-test     5.0  0.2             1.0   
173  google-landmarks  google-landmarks-test     5.0  1.0             0.0   
142  google-landmarks  google-landmarks-test     5.0  1.0             1.0   
133  google-landmarks  google-landmarks-test     5.0  1.0             0.0   
209  google-landmarks  google-landmarks-test     5.0  0.2             1.0   
151  google-landmarks  google-landmarks-test     5.0  0.2             0.0   
227  google-landmarks  google-landmarks-test     5.0  0.2             0.0   
191  google-landmarks  google-landmarks-test     5.0  0.2             0.0   
194  google-landmarks  google-landmarks-test     5.0  0.2             0.0   
160  google-landmarks  google-landmarks-test     5.0  0.2             1.0   
203  google-landmarks  google-landmarks-test     5.0  0.2             1.0   
206  google-landmarks  google-landmarks-test     5.0  0.2             1.0   

     dfmn_k is_feat      distance  test_accuracy  test_accuracy_std  \
176     0.0    True     euclidean       0.403063           0.002916   
154     1.0     NaN     euclidean       0.398087           0.003023   
185     0.0    True     euclidean       0.397973           0.003024   
221     0.0    True          sen2       0.397013           0.002938   
212     0.0   False          sen2       0.394947           0.002881   
215     0.0    True          sen2       0.394810           0.003048   
145     0.0     NaN     euclidean       0.391510           0.003040   
218     0.0   False          sen2       0.389960           0.002931   
136     0.0     NaN     euclidean       0.389737           0.002974   
182     0.0    True  cosine_scale       0.388960           0.002961   
224     1.0   False          sen2       0.388860           0.002988   
200     1.0   False          sen2       0.386233           0.002967   
163     1.0     NaN     euclidean       0.385843           0.002915   
173     0.0    True  cosine_scale       0.385330           0.002956   
142     0.0     NaN  cosine_scale       0.382373           0.002881   
133     0.0     NaN  cosine_scale       0.375750           0.002962   
209     1.0    True     euclidean       0.375747           0.002951   
151     1.0     NaN  cosine_scale       0.375173           0.002951   
227     1.0    True          sen2       0.371747           0.002879   
191     1.0    True  cosine_scale       0.371570           0.002956   
194     1.0    True     euclidean       0.371563           0.002908   
160     1.0     NaN  cosine_scale       0.369180           0.002924   
203     1.0    True          sen2       0.365283           0.002928   
206     1.0    True  cosine_scale       0.361760           0.002853   

     test_top_3_accuracy  test_top_3_accuracy_std  
176             0.564527                 0.003058  
154             0.555110                 0.003080  
185             0.558590                 0.002981  
221             0.557697                 0.002992  
212             0.555990                 0.003006  


* ## miniImageNet -> GoogleLandmarks (100-way)

In [6]:
records_miniImageNet_googleLandmarks = records[(records['dataset'] == 'miniImageNet') & (
    records['test_dataset'] == 'google-landmarks-test')]
records_miniImageNet_googleLandmarks

dataset           test_dataset  n_shot    k  aux_rotation_k  dfmn_k  \
334  miniImageNet  google-landmarks-test     5.0  0.2             0.0     1.0   
346  miniImageNet  google-landmarks-test     5.0  0.2             1.0     1.0   
340  miniImageNet  google-landmarks-test     5.0  0.2             0.0     1.0   
358  miniImageNet  google-landmarks-test     5.0  0.2             1.0     1.0   
352  miniImageNet  google-landmarks-test     5.0  0.2             1.0     1.0   
328  miniImageNet  google-landmarks-test     5.0  0.2             0.0     1.0   
361  miniImageNet  google-landmarks-test     5.0  0.2             1.0     1.0   
355  miniImageNet  google-landmarks-test     5.0  0.2             1.0     1.0   
343  miniImageNet  google-landmarks-test     5.0  0.2             0.0     1.0   
331  miniImageNet  google-landmarks-test     5.0  0.2             0.0     1.0   
337  miniImageNet  google-landmarks-test     5.0  0.2             0.0     1.0   
349  miniImageNet  google-landmarks-test     5.0  0.2             1.0     1.0   
310  miniImageNet  google-landmarks-test     5.0  1.0             1.0     0.0   
319  miniImageNet  google-landmarks-test     5.0  1.0             1.0     0.0   
322  miniImageNet  google-landmarks-test     5.0  1.0             1.0     0.0   
316  miniImageNet  google-landmarks-test     5.0  1.0             1.0     0.0   
313  miniImageNet  google-landmarks-test     5.0  1.0             1.0     0.0   
325  miniImageNet  google-landmarks-test     5.0  1.0             1.0     0.0   
289  miniImageNet  google-landmarks-test     5.0  1.0             0.0     0.0   
295  miniImageNet  google-landmarks-test     5.0  1.0             0.0     0.0   
292  miniImageNet  google-landmarks-test     5.0  1.0             0.0     0.0   
298  miniImageNet  google-landmarks-test     5.0  1.0             0.0     0.0   
301  miniImageNet  google-landmarks-test     5.0  1.0             0.0     0.0   
304  miniImageNet  google-landmarks-test     5.0  1.0             0.0     0.0   
273  miniImageNet  google-landmarks-test     1.0  0.2             0.0     1.0   
267  miniImageNet  google-landmarks-test     1.0  0.2             0.0     1.0   
237  miniImageNet  google-landmarks-test     1.0  0.2             1.0     1.0   
110  miniImageNet  google-landmarks-test     1.0  0.2             1.0     1.0   
264  miniImageNet  google-landmarks-test     1.0  0.2             0.0     1.0   
101  miniImageNet  google-landmarks-test     1.0  0.2             0.0     1.0   
104  miniImageNet  google-landmarks-test     1.0  0.2             0.0     1.0   
230  miniImageNet  google-landmarks-test     1.0  0.2             1.0     1.0   
107  miniImageNet  google-landmarks-test     1.0  0.2             1.0     1.0   
286  miniImageNet  google-landmarks-test     1.0  0.2             1.0     1.0   
233  miniImageNet  google-landmarks-test     1.0  0.2             1.0     1.0   
270  miniImageNet  google-landmarks-test     1.0  0.2             0.0     1.0   
258  miniImageNet  google-landmarks-test     1.0  1.0             1.0     0.0   
255  miniImageNet  google-landmarks-test     1.0  1.0             1.0     0.0   
261  miniImageNet  google-landmarks-test     1.0  1.0             1.0     0.0   
246  miniImageNet  google-landmarks-test     1.0  1.0             0.0     0.0   
249  miniImageNet  google-landmarks-test     1.0  1.0             0.0     0.0   
252  miniImageNet  google-landmarks-test     1.0  1.0             1.0     0.0   
243  miniImageNet  google-landmarks-test     1.0  1.0             0.0     0.0   
98   miniImageNet  google-landmarks-test     1.0  1.0             1.0     0.0   
240  miniImageNet  google-landmarks-test     1.0  1.0             0.0     0.0   
92   miniImageNet  google-landmarks-test     1.0  1.0             0.0     0.0   
95   miniImageNet  google-landmarks-test     1.0  1.0             1.0     0.0   
89   miniImageNet  google-landmarks-test     1.0  1.0             0.0     0.0   

    is_feat      distance  test_accuracy

# Обработанные результаты

### Основа - https://arxiv.org/pdf/1703.05175.pdf - для каждого экземпляра класса получаем вектор признаков, затем усредняем по классам

Реализованы следующие "модули":

* Функции потерь
    1. DFMN - https://arxiv.org/abs/2002.12017v1 - глобальные прототипы для всех классов из тренировочного набора
    2. Rotation task - https://openaccess.thecvf.com/content_ICCV_2019/papers/Gidaris_Boosting_Few-Shot_Visual_Learning_With_Self-Supervision_ICCV_2019_paper.pdf - параллельно с основной задачей учимся распозновать, на какой угол повернули картинку (0, 90, 180 или 270 градусов)
* Преобразования прототипов
    1. FEAT - https://openaccess.thecvf.com/content_CVPR_2020/html/Ye_Few-Shot_Learning_via_Embedding_Adaptation_With_Set-to-Set_Functions_CVPR_2020_paper.html - архитектура "трансформер" - преобразовываем прототип с опорой на остальные прототипы.
* Функции расстояния
    1. Euclidean - обычное евклидово расстояние
    2. Scaled euclidean - https://arxiv.org/abs/2002.12017v1 - евклидово расстояние между нормированными векторами, допольнительно умноженными на число, определяемое для вектора отедльной нейросетью
    3. SEN - https://www.ecva.net/papers/eccv_2020/papers_ECCV/html/4238_ECCV_2020_paper.php - корень из суммы евклидового расстояния между векторами и евклидового расстояния между длинами векторов, умноженным на коэффициент. Во время обучения коэффициент разный для правильных и неправильных пар (1.0 и -(10^-7) соотвественно), во время теста он равен 1.0
    
### Датасеты - miniImageNet - мало разных классов и подмножество googleLandmarks - много похожих классов

Сценарии (1-shot для googleLandmarks не проверялось, тк точность будет слишком маленькая):

* Обучаем на miniImageNet (15-way classification), тестируем на miniImageNet (5-way classification) - стандартный датасет в статьях
    * 1-shot
    * 5-shot
    
* Обучаем на miniImageNet (15-way classification), тестируем на googleLandmarks (100-way classification) - адаптируемся к новому датасету, в котором меньше разнообразие и больше классов
    * 5-shot

* Обучаем на googleLandmarks (15-way classification), тестируем на miniImageNet (5-way classification) - адаптируемся к новому датасету, в котором больше разнообразие и меньше классов
    * 5-shot

* Обучаем на googleLandmarks (15-way classification), тестируем на googleLandmarks (100-way classification)
    * 5-shot  



In [7]:
DISTANCES = {
    'cosine_scale': 'Scaled euclidean',
    'euclidean': 'Euclidean',
    'sen2': 'SEN',
}

def bool_repr(value):
    return '+' if value else '-'

def row2params(row):
    return bool_repr(row['aux_rotation_k'] == 1.0), bool_repr(row['dfmn_k'] == 1.0), bool_repr(row['is_feat'] == True), DISTANCES[row['distance']]


def accuracy_repr(row):
    return "%.3f $+-$ %.3f" % (row['test_accuracy'], row['test_accuracy_std'])


def top3_accuracy_repr(row):
    return "%.3f $+-$ %.3f" % (row['test_top_3_accuracy'], row['test_top_3_accuracy_std'])

In [8]:
def tabulate_bold(*args, **kwargs):
    res = tabulate(*args, **kwargs).split('\n')
    ln = res[4]
    its = ln.split()
    for i in range(len(its)):
        if not (len(its[i]) == 0 or its[i][-1] == '&' or its[i][-1] == '\\' or its[i][-1] == '$'):
            its[i] = '\\textbf{' + its[i] + '}'
    res[4] = ' '.join(its)
    return '\n'.join(res)

In [9]:
configurations = {}

for i, row in records.iterrows():
    ds_from = row['dataset']
    ds_to = row['test_dataset']
    n_shot = str(int(row['n_shot']))

    case = ds_from + '2' + ds_to + '_' + n_shot + 'shot'

    cur = row2params(row)

    if cur not in configurations:
        configurations[cur] = {}

    configurations[cur][case] = {
        'accuracy': row['test_accuracy'],
        'accuracy_std': row['test_accuracy_std'],
        'top3_accuracy': row['test_top_3_accuracy'],
        'top3_accuracy_std': row['test_top_3_accuracy_std'],
        'row': row
    }

configurations

{('-',
  '+',
  '-',
  'Scaled euclidean'): {'miniImageNet2miniImageNet-test_5shot': {'accuracy': 0.7025599999999999,
   'accuracy_std': 0.01113773166612783,
   'top3_accuracy': 0.954559862613678,
   'top3_accuracy_std': 0.00446569122090389,
   'row': dataset                         miniImageNet
   test_dataset               miniImageNet-test
   n_shot                                     5
   k                                        0.2
   aux_rotation_k                             0
   dfmn_k                                     1
   is_feat                                False
   distance                        cosine_scale
   test_accuracy                        0.70256
   test_accuracy_std                  0.0111377
   test_top_3_accuracy                  0.95456
   test_top_3_accuracy_std           0.00446569
   Name: 335, dtype: object}, 'miniImageNet2miniImageNet-test_1shot': {'accuracy': 0.5417200000000002,
   'accuracy_std': 0.01216828013490162,
   'top3_accuracy': 0.8883197903

In [10]:
HEADERS = ['Rotation task', 'DFMN', 'FEAT',
           'Distance', 'Accuracy', 'Top-3 accuracy']

* ## [1-shot] miniImageNet -> miniImageNet (5-way)

In [11]:
CASE = 'miniImageNet2miniImageNet-test_1shot'

cur_configs = []

for config in configurations:
    metrics = configurations[config]
    cur_configs.append(
        (metrics[CASE]['accuracy'], metrics[CASE]['top3_accuracy'], config))

cur_configs.sort(reverse=True)

table = []

for _, _, config in cur_configs:
    row = configurations[config][CASE]['row']
    table.append((*config, accuracy_repr(row), top3_accuracy_repr(row)))

print(tabulate(table, headers=HEADERS, showindex="always", tablefmt="fancy_grid"))

╒════╤═════════════════╤════════╤════════╤══════════════════╤══════════════════╤══════════════════╕
│    │ Rotation task   │ DFMN   │ FEAT   │ Distance         │ Accuracy         │ Top-3 accuracy   │
╞════╪═════════════════╪════════╪════════╪══════════════════╪══════════════════╪══════════════════╡
│  0 │ +               │ -      │ +      │ SEN              │ 0.550 $+-$ 0.013 │ 0.894 $+-$ 0.008 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  1 │ +               │ -      │ +      │ Euclidean        │ 0.549 $+-$ 0.012 │ 0.889 $+-$ 0.008 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  2 │ -               │ -      │ +      │ SEN              │ 0.548 $+-$ 0.012 │ 0.895 $+-$ 0.008 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  3 │ -               │ +      │ +      │ SEN              │ 0.548 $+-$ 0.012 │ 0.894 $+-$ 0.008 │


In [12]:
print(tabulate_bold(table, headers=HEADERS, showindex=range(1, len(table) + 1), tablefmt="latex").replace('+-', '\pm').replace('\\$', '$'))

\begin{tabular}{rllllll}
\hline
    & Rotation task   & DFMN   & FEAT   & Distance         & Accuracy         & Top-3 accuracy   \\
\hline
\textbf{1} & \textbf{+} & \textbf{-} & \textbf{+} & \textbf{SEN} & \textbf{0.550} $\pm$ \textbf{0.013} & \textbf{0.894} $\pm$ \textbf{0.008} \\
  2 & +               & -      & +      & Euclidean        & 0.549 $\pm$ 0.012 & 0.889 $\pm$ 0.008 \\
  3 & -               & -      & +      & SEN              & 0.548 $\pm$ 0.012 & 0.895 $\pm$ 0.008 \\
  4 & -               & +      & +      & SEN              & 0.548 $\pm$ 0.012 & 0.894 $\pm$ 0.008 \\
  5 & -               & +      & +      & Euclidean        & 0.547 $\pm$ 0.013 & 0.889 $\pm$ 0.008 \\
  6 & +               & -      & +      & Scaled euclidean & 0.546 $\pm$ 0.013 & 0.892 $\pm$ 0.008 \\
  7 & -               & -      & +      & Euclidean        & 0.546 $\pm$ 0.013 & 0.893 $\pm$ 0.008 \\
  8 & -               & -      & +      & Scaled euclidean & 0.544 $\pm$ 0.012 & 0.891 $\pm$ 0.008 \\
  9

* ## [5-shot] miniImageNet -> miniImageNet (5-way)

In [13]:
CASE = 'miniImageNet2miniImageNet-test_5shot'

cur_configs = []

for config in configurations:
    metrics = configurations[config]
    cur_configs.append(
        (metrics[CASE]['accuracy'], metrics[CASE]['top3_accuracy'], config))

cur_configs.sort(reverse=True)

table = []

for _, _, config in cur_configs:
    row = configurations[config][CASE]['row']
    table.append((*config, accuracy_repr(row), top3_accuracy_repr(row)))

print(tabulate(table, headers=HEADERS, showindex="always", tablefmt="fancy_grid"))

╒════╤═════════════════╤════════╤════════╤══════════════════╤══════════════════╤══════════════════╕
│    │ Rotation task   │ DFMN   │ FEAT   │ Distance         │ Accuracy         │ Top-3 accuracy   │
╞════╪═════════════════╪════════╪════════╪══════════════════╪══════════════════╪══════════════════╡
│  0 │ -               │ +      │ -      │ Scaled euclidean │ 0.703 $+-$ 0.011 │ 0.955 $+-$ 0.004 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  1 │ +               │ -      │ +      │ Euclidean        │ 0.700 $+-$ 0.011 │ 0.954 $+-$ 0.005 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  2 │ +               │ -      │ +      │ Scaled euclidean │ 0.700 $+-$ 0.011 │ 0.953 $+-$ 0.005 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  3 │ -               │ +      │ -      │ SEN              │ 0.697 $+-$ 0.011 │ 0.950 $+-$ 0.005 │


In [14]:
print(tabulate_bold(table, headers=HEADERS, showindex=range(1, len(table) + 1), tablefmt="latex").replace('+-', '\pm').replace('\\$', '$'))

\begin{tabular}{rllllll}
\hline
    & Rotation task   & DFMN   & FEAT   & Distance         & Accuracy         & Top-3 accuracy   \\
\hline
\textbf{1} & \textbf{-} & \textbf{+} & \textbf{-} & \textbf{Scaled} \textbf{euclidean} & \textbf{0.703} $\pm$ \textbf{0.011} & \textbf{0.955} $\pm$ \textbf{0.004} \\
  2 & +               & -      & +      & Euclidean        & 0.700 $\pm$ 0.011 & 0.954 $\pm$ 0.005 \\
  3 & +               & -      & +      & Scaled euclidean & 0.700 $\pm$ 0.011 & 0.953 $\pm$ 0.005 \\
  4 & -               & +      & -      & SEN              & 0.697 $\pm$ 0.011 & 0.950 $\pm$ 0.005 \\
  5 & -               & +      & -      & Euclidean        & 0.696 $\pm$ 0.011 & 0.951 $\pm$ 0.005 \\
  6 & -               & -      & +      & Euclidean        & 0.694 $\pm$ 0.011 & 0.950 $\pm$ 0.005 \\
  7 & +               & -      & +      & SEN              & 0.693 $\pm$ 0.011 & 0.951 $\pm$ 0.005 \\
  8 & -               & +      & +      & Euclidean        & 0.692 $\pm$ 0.011 & 0.

* ## [5-shot] miniImageNet -> googleLandmarks (100-way)

In [15]:
CASE = 'miniImageNet2google-landmarks-test_5shot'

cur_configs = []

for config in configurations:
    metrics = configurations[config]
    cur_configs.append(
        (metrics[CASE]['accuracy'], metrics[CASE]['top3_accuracy'], config))

cur_configs.sort(reverse=True)

table = []

for _, _, config in cur_configs:
    row = configurations[config][CASE]['row']
    table.append((*config, accuracy_repr(row), top3_accuracy_repr(row)))

print(tabulate(table, headers=HEADERS, showindex="always", tablefmt="fancy_grid"))

╒════╤═════════════════╤════════╤════════╤══════════════════╤══════════════════╤══════════════════╕
│    │ Rotation task   │ DFMN   │ FEAT   │ Distance         │ Accuracy         │ Top-3 accuracy   │
╞════╪═════════════════╪════════╪════════╪══════════════════╪══════════════════╪══════════════════╡
│  0 │ -               │ +      │ -      │ Scaled euclidean │ 0.298 $+-$ 0.003 │ 0.442 $+-$ 0.003 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  1 │ +               │ +      │ -      │ Scaled euclidean │ 0.296 $+-$ 0.003 │ 0.441 $+-$ 0.003 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  2 │ -               │ +      │ -      │ Euclidean        │ 0.295 $+-$ 0.003 │ 0.436 $+-$ 0.003 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  3 │ +               │ +      │ -      │ Euclidean        │ 0.291 $+-$ 0.003 │ 0.434 $+-$ 0.003 │


In [16]:
print(tabulate_bold(table, headers=HEADERS, showindex=range(1, len(table) + 1), tablefmt="latex").replace('+-', '\pm').replace('\\$', '$'))

\begin{tabular}{rllllll}
\hline
    & Rotation task   & DFMN   & FEAT   & Distance         & Accuracy         & Top-3 accuracy   \\
\hline
\textbf{1} & \textbf{-} & \textbf{+} & \textbf{-} & \textbf{Scaled} \textbf{euclidean} & \textbf{0.298} $\pm$ \textbf{0.003} & \textbf{0.442} $\pm$ \textbf{0.003} \\
  2 & +               & +      & -      & Scaled euclidean & 0.296 $\pm$ 0.003 & 0.441 $\pm$ 0.003 \\
  3 & -               & +      & -      & Euclidean        & 0.295 $\pm$ 0.003 & 0.436 $\pm$ 0.003 \\
  4 & +               & +      & -      & Euclidean        & 0.291 $\pm$ 0.003 & 0.434 $\pm$ 0.003 \\
  5 & +               & +      & -      & SEN              & 0.291 $\pm$ 0.003 & 0.433 $\pm$ 0.003 \\
  6 & -               & +      & -      & SEN              & 0.288 $\pm$ 0.003 & 0.432 $\pm$ 0.003 \\
  7 & +               & +      & +      & Euclidean        & 0.285 $\pm$ 0.003 & 0.428 $\pm$ 0.003 \\
  8 & +               & +      & +      & SEN              & 0.285 $\pm$ 0.003 & 0.

* ## [5-shot] googleLandmarks -> miniImageNet (5-way)

In [17]:
CASE = 'google-landmarks2miniImageNet-test_5shot'

cur_configs = []

for config in configurations:
    metrics = configurations[config]
    cur_configs.append(
        (metrics[CASE]['accuracy'], metrics[CASE]['top3_accuracy'], config))

cur_configs.sort(reverse=True)

table = []

for _, _, config in cur_configs:
    row = configurations[config][CASE]['row']
    table.append((*config, accuracy_repr(row), top3_accuracy_repr(row)))

print(tabulate(table, headers=HEADERS, showindex="always", tablefmt="fancy_grid"))

╒════╤═════════════════╤════════╤════════╤══════════════════╤══════════════════╤══════════════════╕
│    │ Rotation task   │ DFMN   │ FEAT   │ Distance         │ Accuracy         │ Top-3 accuracy   │
╞════╪═════════════════╪════════╪════════╪══════════════════╪══════════════════╪══════════════════╡
│  0 │ +               │ -      │ +      │ Euclidean        │ 0.562 $+-$ 0.011 │ 0.897 $+-$ 0.007 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  1 │ -               │ -      │ +      │ Euclidean        │ 0.562 $+-$ 0.011 │ 0.896 $+-$ 0.007 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  2 │ +               │ -      │ -      │ SEN              │ 0.561 $+-$ 0.012 │ 0.897 $+-$ 0.007 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  3 │ -               │ -      │ -      │ Euclidean        │ 0.561 $+-$ 0.011 │ 0.895 $+-$ 0.007 │


In [18]:
print(tabulate_bold(table, headers=HEADERS, showindex=range(1, len(table) + 1), tablefmt="latex").replace('+-', '\pm').replace('\\$', '$'))

\begin{tabular}{rllllll}
\hline
    & Rotation task   & DFMN   & FEAT   & Distance         & Accuracy         & Top-3 accuracy   \\
\hline
\textbf{1} & \textbf{+} & \textbf{-} & \textbf{+} & \textbf{Euclidean} & \textbf{0.562} $\pm$ \textbf{0.011} & \textbf{0.897} $\pm$ \textbf{0.007} \\
  2 & -               & -      & +      & Euclidean        & 0.562 $\pm$ 0.011 & 0.896 $\pm$ 0.007 \\
  3 & +               & -      & -      & SEN              & 0.561 $\pm$ 0.012 & 0.897 $\pm$ 0.007 \\
  4 & -               & -      & -      & Euclidean        & 0.561 $\pm$ 0.011 & 0.895 $\pm$ 0.007 \\
  5 & +               & +      & -      & SEN              & 0.558 $\pm$ 0.012 & 0.894 $\pm$ 0.007 \\
  6 & -               & +      & -      & Euclidean        & 0.557 $\pm$ 0.012 & 0.891 $\pm$ 0.007 \\
  7 & +               & -      & -      & Euclidean        & 0.557 $\pm$ 0.012 & 0.898 $\pm$ 0.007 \\
  8 & +               & -      & +      & SEN              & 0.556 $\pm$ 0.011 & 0.891 $\pm$ 0.007 

* ## [5-shot] googleLandmarks -> googleLandmarks (100-way)

In [19]:
CASE = 'google-landmarks2google-landmarks-test_5shot'

cur_configs = []

for config in configurations:
    metrics = configurations[config]
    cur_configs.append(
        (metrics[CASE]['accuracy'], metrics[CASE]['top3_accuracy'], config))

cur_configs.sort(reverse=True)

table = []

for _, _, config in cur_configs:
    row = configurations[config][CASE]['row']
    table.append((*config, accuracy_repr(row), top3_accuracy_repr(row)))

print(tabulate(table, headers=HEADERS, showindex="always", tablefmt="fancy_grid"))

╒════╤═════════════════╤════════╤════════╤══════════════════╤══════════════════╤══════════════════╕
│    │ Rotation task   │ DFMN   │ FEAT   │ Distance         │ Accuracy         │ Top-3 accuracy   │
╞════╪═════════════════╪════════╪════════╪══════════════════╪══════════════════╪══════════════════╡
│  0 │ -               │ -      │ +      │ Euclidean        │ 0.403 $+-$ 0.003 │ 0.565 $+-$ 0.003 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  1 │ -               │ +      │ -      │ Euclidean        │ 0.398 $+-$ 0.003 │ 0.555 $+-$ 0.003 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  2 │ +               │ -      │ +      │ Euclidean        │ 0.398 $+-$ 0.003 │ 0.559 $+-$ 0.003 │
├────┼─────────────────┼────────┼────────┼──────────────────┼──────────────────┼──────────────────┤
│  3 │ +               │ -      │ +      │ SEN              │ 0.397 $+-$ 0.003 │ 0.558 $+-$ 0.003 │


In [20]:
print(tabulate_bold(table, headers=HEADERS, showindex=range(1, len(table) + 1), tablefmt="latex").replace('+-', '\pm').replace('\\$', '$'))

\begin{tabular}{rllllll}
\hline
    & Rotation task   & DFMN   & FEAT   & Distance         & Accuracy         & Top-3 accuracy   \\
\hline
\textbf{1} & \textbf{-} & \textbf{-} & \textbf{+} & \textbf{Euclidean} & \textbf{0.403} $\pm$ \textbf{0.003} & \textbf{0.565} $\pm$ \textbf{0.003} \\
  2 & -               & +      & -      & Euclidean        & 0.398 $\pm$ 0.003 & 0.555 $\pm$ 0.003 \\
  3 & +               & -      & +      & Euclidean        & 0.398 $\pm$ 0.003 & 0.559 $\pm$ 0.003 \\
  4 & +               & -      & +      & SEN              & 0.397 $\pm$ 0.003 & 0.558 $\pm$ 0.003 \\
  5 & -               & -      & -      & SEN              & 0.395 $\pm$ 0.003 & 0.556 $\pm$ 0.003 \\
  6 & -               & -      & +      & SEN              & 0.395 $\pm$ 0.003 & 0.558 $\pm$ 0.003 \\
  7 & +               & -      & -      & Euclidean        & 0.392 $\pm$ 0.003 & 0.551 $\pm$ 0.003 \\
  8 & +               & -      & -      & SEN              & 0.390 $\pm$ 0.003 & 0.549 $\pm$ 0.003 